### ISSUE TITLE
Expid command refactored. - [merged]

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1803]
In GitLab by @dbeltrankyl on Feb 14, 2023, 13:12

_Merges Custom_config_upgrades -> master_

(non-related) Contains, -rm renamed to -rom ( run_members ) 

* Contains: 
  * Expdef full refactor. It was a big mess
  * Added -min, --minimal_config
    * Added -repo, --git_repo
    * Added -branch, --git_branch, default main
    * Added -config, --git_as_conf, relative to project root
  * Reworked dummy
    * Default, hpc = "local", will generate an experiment without comments
  * Reworked default config
    * [TODO] Will generate the same configuration files as dummy but commented
  * Added to autosubmitconfigparser/files
    * Expdef-minimal.yml
    * expdef,autosubmit,jobs, platforms -dummy

Wip, adding #950

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 14, 2023, 14:49

Hello @mcastril  

I'll reply to https://earth.bsc.es/gitlab/digital-twins/de_340/workflow/-/merge_requests/3#note_199044 


> Things to discuss with @dbeltrankyl :

> It would be nice if we could combine CUSTOM_CONFIG set in different files in a nested way. Maybe a way to >overload parameters appending the new value to the original one.
>    About the precedence of the mechanism:

My new plan:


In this branch, -min is already working, but with the name expid.yml, what if I perform the following changes: 

Change the name of expid.yml ( minimal one) to config-minimal.yml

> CUSTOM_CONFIG_USER_FILES > CUSTOM_CONFIG_FOLDER > $EXPID/conf

Change loading order to:

config-minimal.yml > $EXPID/conf > CUSTOM_CONFIG

>It would be nice if we could combine CUSTOM_CONFIG set in different files in a nested way. Maybe a way to overload parameters appending the new value to the original one.

Change Custom_config directive from str to dict

```
Custom_Config:
        order: folder_1,folder_0 # default is dict order
	Folder_0:
		path: <absolute_path>
                order: file_1,file_0 # default is by specified order 
                file_0: <relative_to_path>
	        file_1: <relative_to_path>
	Folder_1: 
		path: <absolute_path>
		order: # default is by file_name 
```

* If only the folder path is specified, all files will be loaded. 
* If you select files, only these files will be loaded.

* With "order" attr, you can change the order from any file. 

* If folder_key exists, it will be overwritten by the new values
* If folder_key doesn't exist, it will be append

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 14, 2023, 16:58

Hi @dbeltrankyl 

> Draft: Expdef refactored. 

Do you mean expid refactored? The branch is about the expid command right? At least this was my first impression by looking at the parameters.

> Change the name of expid.yml ( minimal one) to config-minimal.yml

Here, maybe you mean expdef.yml ? Sorry, I didn't know that there was such file (expid.yml) but maybe this change in AS4.

> Change Custom_config directive from str to dict

I understand that this can be less intrusive that having an additional operator for concatenating and not overloading isn't it?

> folder_key

Folder_Key is a key? Or you mean Folder_# ? I guess it only would be overwritten depending on its position in the order value right?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 15, 2023, 09:55

> Do you mean expid refactored? The branch is about the expid command right? At least this was my first impression by looking at the parameters.

Yes, I just swapped the names, sorry. 

> Here, maybe you mean expdef.yml ? Sorry, I didn't know that there was such file (expid.yml) but maybe this change in AS4.

Yes, again, I swapped the names. 


> I understand that this can be less intrusive than having an additional operator for concatenating and not overloading, isn't it?

The problem with having simple str is that it is hard to know what to load from Autosubmit if multiple files interact with the value. 

> Folder_Key is a key? Or you mean Folder\_# ? I guess it only would be overwritten depending on its position in the order value, right?

It is a key that can be named whatever.


Current (new) plan and default behavior:

1) First, autosubmit, will look at the new minimal-configuration.yml ( or whatever the final name of the file is)
2) it will load the folders/files specified on there and also will add the section keyword to an internal list of "loaded_sections"
3) After all files/folders are loaded, Autosubmit will see if there are changes in the Custom_config and load the sections not listed in "loaded_sections", (repeat 2)

Possible issue:

What if, in the newly loaded files, there are set multiple Custom_config? Which would be the correct order? For Autosubmit, it will come ordered by file_name 

I propose having an optional order keyword for these cases, which would be in a specific.yml(config_order.yml) file inside any current folders to load ( it will not look in the already loaded ones listed in the list of "loaded_sections"). Autosubmit will read this file first to know what to load first. (there would be an issue if config_order.yml is present in multiple sections loaded at the same time tho)

4) Finally, it will load the $expid/conf all *.yml or *.yaml files (Except minimal-configuration.yml). 


Also, I'm thinking of dropping some definition layers.

```
Custom_Config:
        order: file_0,folder_0 # default is append.
	Folder_0: <absolute_path> # can be relative to project_root if it doesn't start with "/"
	file_0: <absolute_path> # can be a file
```

To simplify the complexity, however, it means that all files in a folder will always be loaded.


In conclusion, hierchancy in custom_config can be a bit tricky

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 15, 2023, 17:50

> 1. First, autosubmit, will look at the new minimal-configuration.yml ( or whatever the final name of the file is)
> 1. it will load the folders/files specified on there and also will add the section keyword to an internal list of "loaded_sections"
> 1. After all files/folders are loaded, Autosubmit will see if there are changes in the Custom_config and load the sections not listed in "loaded_sections", (repeat 2)

So it seems to me that the `Custom_config` section will be progressively loaded in a recursive manner right?

Maybe we can simplify the order as here.

If A is the minimal configuration in $EXPID/conf, and it includes different entries (B, C) in CUSTOM_CONFIG. And if C contains another CUSTOM_CONFIG entry pointing at D:

A -> B  
A -> C -> D  
A -> E  

Autosubmit would first load B (without loading the rest of the configuration in A). Then we proceed to the next section, and C would overload any parameter in B, and D any parameter on those. Finally, E would overload anything read above.

If A is the minimal config, B a template config in the project (could be the one having the most general information), C and D additional project templates (with model or particular configuration information) and E is the user config, I think that we fulfill the purpose.

The user only has to remind to load the configs in the appropriate order. I would not make any distinction between directories and files to make it simpler. Do you foresee any problem with this configuration?

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 15, 2023, 17:57

Thinking again, I think this is now enough.

* I forgot to say that after loading all the CUSTOM_CONFIG, then Autosubmit would proceed to read the rest of parameters in A (out of CUSTOM_CONFIG). 
* Then, we only have the problem of how overloading parameters loaded in a local configuration ($EXPID/conf) that is not minimal.

We can add a parameter to CUSTOM_CONFIG to indicate when to load the configuration?

Simple:

```
CUSTOM_CONFIG: "my_first_directory/, my_second_directory/, my_file"
```


Advanced:

```
CUSTOM_CONFIG: 
* PRE: "my_first_directory/, my_second_directory/, my_file"
* POST: "my_file_in_home"
```

* All directories/files in PRE would be loaded before the rest of parameters in the main config, and overloaded by the main config.
* All directories/files in POST would be loaded after all the parameters in the main config, and would overload those.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 16, 2023, 08:45

> If A is the minimal configuration in $EXPID/conf, and it includes different entries (B, C) in CUSTOM_CONFIG. And if C contains another CUSTOM_CONFIG entry pointing at D:

A -> B  
A -> C -> D  
A -> E

Yes, this was the idea. But it only works properly if a single CUSTOM_CONFIG directive is specified per folder.

> And if C contains another CUSTOM_CONFIG entry pointing at D:

My problem was if, for example, B has multiple CUSTOM_CONFIG entries across the folder(in multiple files). Autosubmit would recursively load the first one encountered (A->B1->F->G, A->B2->H-I). But what if the user wanted to load first B2? for that reason, I thought about the order keyword that was intended to use in the scope of the folder loaded.


Another "solution" would be to raise an error if multiple custom_config are detected across the files in the same folder.

> Possible issue:

> What if, in the newly loaded files, there are set multiple Custom_config? Which would be the correct order? For Autosubmit, it will come ordered by file_name
> 
> I propose having an optional order keyword for these cases, which would be in a specific.yml(config_order.yml) file inside any current folders to load ( it will not look in the already loaded ones listed in the list of "loaded_sections"). Autosubmit will read this file first to know what to load first. (there would be an issue if config_order.yml is present in multiple sections loaded at the same time tho)

The idea was to temporally load the whole folder without recursivity ( A -> B{B1,B2}) to merge all  possible custom_config directives from a folder, then As would look at the final folder order with the order keyword and load all parameters following the intended hierarchy.

So modifying your example, taking into account multiple custom_config in the same folder ( called B ), AS would do:


* A -> B
  * Default: 
    * B1->F->G 
    * B2->H->I
  * Using order keyword: B1,B2: 
    * B1->F->G 
    * B2->H->I
  * Using order keyword: B2,B1: 
    * B2->H->I
    * B1->F->G
* A -> C -> D 
* A -> E

Detailed example for B case:

* B{folder) -> marenostrum4/ 
  * B1-> "general-config.yml" `CUSTOM_CONFIG: PATH: "F.yml",G.yml" `
  * B2-> "branch-config.yml" `CUSTOM_CONFIG:  PATH:"H.yml,I.yml" ` 
  * B3-> (config-order.yml)  `CUSTOM_CONFIG: ORDER:general-config,branch-config` 

Then for autosubmit, 

* Temporary loads all B files or only B3 ( if we hardcode the filename)
* Looks if "order" is filled, ignoring "path" value
* If filled, recursively reload files in the proper order and add them to the experiment_data
* If not filled, recursively reload files in the default order and add them to the experiment_data.
The only constraint would be having "ORDER" declared once per folder. 

In this case, without order defined, it would load the first branch-config and then general-config


> I forgot to say that after loading all the CUSTOM_CONFIG, then Autosubmit would proceed to read the rest of the parameters in A (out of CUSTOM_CONFIG).

> * Then, we only have the problem of overloading parameters loaded in a local configuration ($EXPID/conf) that is not minimal.

I think this is not a problem. I have to load all $expid/conf in a temporal_dict at the start to look for custom_config directives and start the process, if there aren't CUSTOM_CONFIG, it will just assign this temporal_dict to the final data. If there are CUSTOM_CONFIG, the temporal_dict will be discarded and begin to load from the one on the CUSTOM_CONFIG.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 16, 2023, 08:59

> Simple:
> 
> ```
> CUSTOM_CONFIG: "my_first_directory/, my_second_directory/, my_file"
> ```
> 

I would add a yaml-format alternative to simple. ( although we have still to discuss the order thing ) 

Simple-yaml-format:

```
CUSTOM_CONFIG: 
  - "my_first_directory/"
  - "my_second_directory/"
  - "my_file"
```

> Advanced:
> 
> ```
> CUSTOM_CONFIG: 
> * PRE: "my_first_directory/, my_second_directory/, my_file"
> * POST: "my_file_in_home"
> ```

I didn't explain the user settings with the proposed solution yesterday. I like this approach, I was thinking about adding a sign to the string itself, but I think it would be less messy to do it with these two sections.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 16, 2023, 09:44

added 2 commits

<ul><li>b3b45de6 - min rst</li><li>73f6e1d7 - [wip] config changes</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12460&start_sha=19a9105934bbf3f43a8d351af8faecf96779dba5)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 16, 2023, 10:46

Also, I have to take into account when to reload the files. 

I guess the best solution is to have a list backtracking that stores all filenames stored in order. Then in every autosubmit iteration, autosubmit should see if there are modifications in any of the files it has loaded. 

In case of modification:

I guess the best idea is to reload all files, not just that one because the custom_conf could have changed, until now we only reload the file modified but may be not enough

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 16, 2023, 14:59

```
CUSTOM_CONFIG: 
* PRE: "my_first_directory/, my_second_directory/, my_file"
* POST: "my_file_in_home"
```

@mcastril  Pre and post only make sense if it is set in the minimal-config.yml, no? as this CUSTOM_CONFIG value will change between files, so if it is defined in between, it would be overridden.


This will be the last syntax, no? 
```
CUSTOM_CONFIG: 
* PRE: 
   - "my_first_directory"
   - "my_second_directory"
   - "my_file"
* POST: 
  - "my_file_in_home"
```

and for files in between:

```
CUSTOM_CONFIG: 
   - "my_first_directory"
   - "my_second_directory"
   - "my_file"
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 16, 2023, 18:41

> I think this is not a problem. I have to load all $expid/conf in a temporal_dict at the start to look for custom_config directives and start the process, if there aren't CUSTOM_CONFIG, it will just assign this temporal_dict to the final data. If there are CUSTOM_CONFIG, the temporal_dict will be discarded and begin to load from the one on the CUSTOM_CONFIG.

Thanks @dbeltrankyl. My point was that, with an order or PRE/POST directive, if the conf/ files are the last ones to be loaded (which is desirable so the user can locally overload parameters), then there is no way of delegating some of the parameters (as the platform) to some file loaded at the end. But it could be solved with the POST.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 16, 2023, 18:43

> I didn't explain the user settings with the proposed solution yesterday. I like this approach, I was thinking about adding a sign to the string itself, but I think it would be less messy to do it with these two sections.

I'm glad that you find it less messy. I think this gives an easy way that users can understand without too much trouble.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 16, 2023, 18:46

> @mcastril Pre and post only make sense if it is set in the minimal-config.yml, no? as this CUSTOM_CONFIG value will change between files, so if it is defined in between, it would be overridden.

Do you mean that in the `minimal-config.yml` is not expected to have recursion when loading nested CUSTOM_CONFIGs?

I don't expect that we use PRE/POST out of the local `/conf`, but in case that we do that, I think it can work the same (load directories in POST after loading the PRE and the rest of the config parameters)

You mean that CUSTOM_CONFIG would be overridden when you load a file with a CUSTOM_CONFIG key? I didn't expect this to happen, because it would happen the same no matter if we use PRE/POST or not.

Maybe I didn't understand well the issue

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 21, 2023, 15:35

added 1 commit

<ul><li>9d5a7f59 - Still some bugs</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12567&start_sha=73f6e1d73119cf68670c04635203dde0d55e03a0)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 22, 2023, 14:51

added 1 commit

<ul><li>f81d19d3 - Recursive config working</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12579&start_sha=9d5a7f591c8364c94f65b0a0b20e1bf0024ce3fa)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 22, 2023, 15:47

added 1 commit

<ul><li>a7240015 - Now autosubmit will ignore -minimal files.</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12584&start_sha=f81d19d367743218d3d28c42232ebd8527c13746)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 22, 2023, 16:18

Ok. I think it is working now. I have to do more tests but so far, so good. 


### Unavoidable Issues:
  * Files ending with "minimal.yml" in the $expid/conf folder will be ignored. Otherwise, they would override the "PRE" values.
  * When using autosubmit -v, if the "AUTOSUBMIT.VERSION" key is not found, Autosubmit will generate a $expid/"version.yml". ( to store the -v )
  * Autosubmit can no longer modify physical files as it doesn't know where they are. 

* Only one **custom_config directive** per **folder**, but you can load file by file. 

* A file can only be loaded **ONCE**.

### Changes:

* There are no  names constraint, except for the expid-minimal.yml (the expid generates that file via -min parameter)

* Now, Autosubmit can load custom_data from custom_data. Previously it only allowed one level of distributed data.


### Custom_data order: 

For each file/folder specified in custom_config of the current folder/file: 

* custom_config: PRE files in recursive order.
* current_folder_data without additional custom_config data
* custon_config: POST files in recursive order.

### Minimal order: 

PRE=PROJ_DATA -> $EXPID/CONF -> POST=USER_DATA

### Final order: 

PROJ/DATA -> $EXPID/CONF DATA -> USER/DATA

### How to achieve user configuration: 

Just add a POST key in the minimal.yml file. This will have a higher priority.

### Example of a minimal configuration:

```
CONFIG:
  AUTOSUBMIT_VERSION: "4.0.0b"
DEFAULT:
    EXPID: "a065"
    HPCARCH: "local"
    CUSTOM_CONFIG:
      PRE: "%ROOTDIR%/proj/git_project/as_conf" # Loads project data
      POST: "/home/dbeltran/platforms.yml" # loads user data
PROJECT:
    PROJECT_TYPE: "git"
    PROJECT_DESTINATION: "git_project"
GIT:
    PROJECT_ORIGIN: "https://earth.bsc.es/gitlab/ces/auto-advanced_config_example"
    PROJECT_BRANCH: "main"
    PROJECT_COMMIT: ""
    PROJECT_SUBMODULES: ""
    FETCH_SINGLE_BRANCH: True
```

### Known issues: 

* Reload is now reloading all files if ANY file has been modified. Previously, it only was the file moded. Knowing what else to load is not trivial if a file is modded, as it can affect the rest of the configuration. Anyway, its impact is mitigated as this only happens if the user modifies the file. 


### In summary, the user only has to know that:

* Pre/POST inside proj data or user data can be used to define the order
* PRE inside minimal-conf should point to project data. 
* POST inside minimal-conf should point to the personal user data.
* A file is only loaded once. It will be ignored if multiple custom_config points to the same file. 
* A folder should have only one custom_config directive.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 24, 2023, 17:32

mentioned in merge request ces/autosubmit4-config-parser!1

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 24, 2023, 17:32

marked this merge request as **ready**

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 24, 2023, 17:32

requested review from @kinow

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 24, 2023, 17:33

> Only one **custom_config directive** per **folder**, but you can load file by file.

I did changes to solve this, now this is allowed

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 24, 2023, 17:34

I think it is ready now, I've to add some related changes, but they are being done in another branch. (: edit had some issues but rebase done) 
Also, I need to rewrite the rst page since the CUSTOM_CONFIG priority has changed. Also, I will clarify some points, like the folder inside folders aren't being loaded, only the YAML files of that specific folder

Please also check https://earth.bsc.es/gitlab/ces/autosubmit4-config-parser/-/merge_requests/1 as the main changes are on there

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 24, 2023, 18:11

added 43 commits

<ul><li>a7240015...aace6b4c - 31 commits from branch <code>master</code></li><li>073889fb - resolves #829,</li><li>64885968 - Expdef refactored.</li><li>94bea0f1 - -Copy working</li><li>04afb7af - -min working in expdef</li><li>e8843e77 - -min working</li><li>cbb1b630 - min rst</li><li>225bd014 - [wip] config changes</li><li>e5e9e49d - Still some bugs</li><li>26385472 - Recursive config working</li><li>17f37deb - Now autosubmit will ignore -minimal files.</li><li>163cae97 - rebase and typo in function name</li><li>221d8a33 - merge conflict</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12679&start_sha=a7240015c662f3af252c6179191fc454bf696de7)

**Comment by kinow** [comment]
Reviewing this one today :thumbsup:

**Comment by kinow** [comment]
`-config/--git_as_conf` is missing from the usage.

**Comment by kinow** [comment]
s/branch/git_branch

**Comment by kinow** [comment]
s/config_path/git_as_conf ?

**Comment by kinow** [comment]
No `test` arg, and I think it's missing `git_branch`?

**Comment by kinow** [comment]
Finished an initial review of this one, also saw that the build is failing due to this, @dbeltrankyl 

```bash
======================================================================
ERROR: test_job_script_checking_contains_the_right_default_variables (test.unit.test_job.TestJob)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gitlab-runner/builds/o7zBmX1g/1/gitlab/es/autosubmit/test/unit/test_job.py", line 294, in test_job_script_checking_contains_the_right_default_variables
    parameters = self.job.update_parameters(self.as_conf, dict())
  File "/home/gitlab-runner/builds/o7zBmX1g/1/gitlab/es/autosubmit/autosubmit/job/job.py", line 1001, in update_parameters
    parameters['JOBNAME'] = self.name
TypeError: 'Mock' object does not support item assignment
Name   
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 14:50

added 1 commit

<ul><li>21222186 - Refactored parameters function, added nodes tasks processors and threads at platform level</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12730&start_sha=221d8a33532b6f98cb8eef761211740dc26b9ced)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 14:54

Yes, the pipeline will fail because it should rely on autosubmitconfigparser 1.0.0, which is not yet released. 

I'll have to fix the pipeline on master or in another branch after the merge

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 14:59

Commented on [docs/source/userguide/run/index.rst line 31](https://github.com/BSC-ES/autosubmit/compare/aace6b4ce6f5ee888cc07f10d170236efb95ac70..f82bfd2ade86420edbc34d7f3eed7a178c6cb514#diff-4e64aab175be3a8bd3ac56c25c1bc9daR31)



We had one, but we switched to the wiki

https://earth.bsc.es/gitlab/es/autosubmit/-/wikis/versions

There is no changelog done for 4.0.0 tho, there is the RST page that list the mayor changes but it is not being updated since a while

**Comment by kinow** [comment]
>Yes, the pipeline will fail because it should rely on autosubmitconfigparser 1.0.0, which is not yet released.
>
>I'll have to fix the pipeline on master or in another branch after the merge

:thumbsup:

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 16:02

added 1 commit

<ul><li>30e37f29 - This should fix the pipeline, also fixed an possible issue with EXPORT...</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12736&start_sha=2122218629713d05f9d6b3cbf9fee81ac35c753d)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 16:11

seems resolved now

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 16:14

Commented on [autosubmit/autosubmit.py line 1103](https://github.com/BSC-ES/autosubmit/compare/aace6b4ce6f5ee888cc07f10d170236efb95ac70..30e37f29c4b38e4b81f1e13d385a55d582743983#diff-fb87ba4e4225b357016f4298108bcf0dR1103)



changed this line in [version 10 of the diff](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12738&start_sha=30e37f29c4b38e4b81f1e13d385a55d582743983#6bde465286156a9da245a31e64df703a55c76323_1103_1103)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 16:14

added 1 commit

<ul><li>6c7080c0 - git test,git_branch cleaned expdef-minimal for unused parameter</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12738&start_sha=30e37f29c4b38e4b81f1e13d385a55d582743983)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 16:14

Commented on [autosubmit/autosubmit.py line 1103](https://github.com/BSC-ES/autosubmit/compare/aace6b4ce6f5ee888cc07f10d170236efb95ac70..30e37f29c4b38e4b81f1e13d385a55d582743983#diff-fb87ba4e4225b357016f4298108bcf0dR1103)



Yes, solved

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 16:19

Commented on [autosubmit/autosubmit.py line 1054](https://github.com/BSC-ES/autosubmit/compare/aace6b4ce6f5ee888cc07f10d170236efb95ac70..6c7080c03c749b67e29aebbb20288dff7b1c5735#diff-fb87ba4e4225b357016f4298108bcf0dR1054)



changed this line in [version 11 of the diff](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12740&start_sha=6c7080c03c749b67e29aebbb20288dff7b1c5735#6bde465286156a9da245a31e64df703a55c76323_1054_1054)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 16:19

Commented on [autosubmit/autosubmit.py line 1055](https://github.com/BSC-ES/autosubmit/compare/aace6b4ce6f5ee888cc07f10d170236efb95ac70..6c7080c03c749b67e29aebbb20288dff7b1c5735#diff-fb87ba4e4225b357016f4298108bcf0dR1055)



changed this line in [version 11 of the diff](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12740&start_sha=6c7080c03c749b67e29aebbb20288dff7b1c5735#6bde465286156a9da245a31e64df703a55c76323_1055_1054)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 16:19

added 1 commit

<ul><li>9a5172e1 - more docstrings</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12740&start_sha=6c7080c03c749b67e29aebbb20288dff7b1c5735)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 16:19

Commented on [autosubmit/autosubmit.py line 1055](https://github.com/BSC-ES/autosubmit/compare/aace6b4ce6f5ee888cc07f10d170236efb95ac70..6c7080c03c749b67e29aebbb20288dff7b1c5735#diff-fb87ba4e4225b357016f4298108bcf0dR1055)



Fixed

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 28, 2023, 16:31

Commented on [autosubmit/autosubmit.py line 1044](https://github.com/BSC-ES/autosubmit/compare/aace6b4ce6f5ee888cc07f10d170236efb95ac70..f82bfd2ade86420edbc34d7f3eed7a178c6cb514#diff-fb87ba4e4225b357016f4298108bcf0dR1044)



Well, the minimal.yml comes from the autosubmitconfigparser4/files/expid-minimal.yml so it won't have another extension, but you're right I'll add the yaml just in case

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 1, 2023, 10:07

Commented on [autosubmit/autosubmit.py line 1021](https://github.com/BSC-ES/autosubmit/compare/aace6b4ce6f5ee888cc07f10d170236efb95ac70..9a5172e1054e9be7a31a9471dc7b4ff092e16233#diff-fb87ba4e4225b357016f4298108bcf0dR1021)



changed this line in [version 12 of the diff](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12748&start_sha=9a5172e1054e9be7a31a9471dc7b4ff092e16233#6bde465286156a9da245a31e64df703a55c76323_1021_1021)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 1, 2023, 10:07

added 1 commit

<ul><li>f82bfd2a - simplified expression</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/315/diffs?diff_id=12748&start_sha=9a5172e1054e9be7a31a9471dc7b4ff092e16233)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 1, 2023, 10:08

I think all are addressed, I will merge it to master to start the tests and open a new merge to add docs

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 1, 2023, 10:12

enabled an automatic merge when the pipeline for f82bfd2ade86420edbc34d7f3eed7a178c6cb514 succeeds

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 1, 2023, 10:15

mentioned in commit 84bc7020064a9b54b70a7f86b7b187cdac36c525

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 1, 2023, 12:59

> ### In summary, the user only has to know that:
> 
> * Pre/POST inside proj data or user data can be used to define the order
> * PRE inside minimal-conf should point to project data.
> * POST inside minimal-conf should point to the personal user data.
> * A file is only loaded once. It will be ignored if multiple custom_config points to the same file.
> * A folder should have only one custom_config directive.

Thanks Dani for explaining so nicely. This is important information that should be added to the docs.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 1, 2023, 12:59

Perfect!

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 1, 2023, 13:04

Commented on [docs/source/userguide/run/index.rst line 31](https://github.com/BSC-ES/autosubmit/compare/aace6b4ce6f5ee888cc07f10d170236efb95ac70..f82bfd2ade86420edbc34d7f3eed7a178c6cb514#diff-4e64aab175be3a8bd3ac56c25c1bc9daR31)



Now that we become more open I think that we need something at the general readthedocs.
